src/bentoml/_internal/service/openapi/utils.py

```diff
@@ -34,6 +69,682 @@
+        return {encode_default(k): encode_default(v) for k, v in o.items()}
+    elif isinstance(o, Enum):
+        return o.value
+    elif isinstance(o, (int, float, str)):
+        return o
+    elif LazyType[ListAny](list).isinstance(o) or LazyType[TupleAny](tuple).isinstance(
+        o
+    ):
+        tp = o.__class__
+        seq_args = (encode_default(i) for i in o)
+        return tp(*seq_args) if is_namedtuple(tp) else tp(seq_args)
+    elif o is None:
+        return None
+    else:
+        return attrs_encoder(o)
+
+
+def get_long_model_name(model: TypeAttrsOrEnum) -> str:
+    return f"{model.__module__}__{model.__qualname__}".replace(".", "__")
+
+
+def normalize_name(name: str) -> str:
+    """
+    Normalizes the given name. This can be applied to either a model *or* enum.
+    """
+    return re.sub(r"[^a-zA-Z0-9.\-_]", "_", name)
+
+
+def get_flat_attrs_from_attr_model(
+    attrs_model: type[AttrsInstance], known_attrs: TypeAttrsSet | None = None
+) -> TypeAttrsSet:
+    known_attrs = known_attrs or set()
+    flat_attrs: TypeAttrsSet = set()
+    flat_attrs.add(attrs_model)
+    known_attrs |= flat_attrs
+    flat_attrs |= get_flat_attrs_from_fields(
+        attr.fields(attrs_model), known_attrs=known_attrs
+    )
+    return flat_attrs
+
+
+def get_flat_attrs_from_fields(
+    fields: tuple[attr.Attribute[t.Any]], known_attrs: TypeAttrsSet
+) -> TypeAttrsSet:
+    flat_models: TypeAttrsSet = set()
+    for field in fields:
+        flat_models |= get_flat_attrs_from_field(field, known_attrs=known_attrs)
+    return flat_models
+
+
+def get_flat_attrs_from_field(
+    field: attr.Attribute[t.Any], known_attrs: TypeAttrsSet
+) -> TypeAttrsSet:
+    flat_models: TypeAttrsSet = set()
+    field_type = field.type
+    if field_type and hasattr(field_type, "type") and attr.has(field_type.type):
+        field_type = field_type.type
+
+    if field_type and attr.has(field_type) and field.type not in known_attrs:
+        flat_models |= get_flat_attrs_from_attr_model(
+            field_type, known_attrs=known_attrs
+        )
+    elif field_type and lenient_issubclass(field_type, Enum):
+        flat_models.add(field_type)
+    return flat_models
+
+
+def get_attrs_name_map(
+    unique_attrs: TypeAttrsSet,
+) -> dict[TypeAttrsOrEnum, str]:
+    name_model_map: dict[str, TypeAttrsOrEnum] = {}
+    conflicts: set[str] = set()
+    for attr_model in unique_attrs:
+        attr_name = normalize_name(attr_model.__name__)
+        if attr_name in conflicts:
+            attr_name = get_long_model_name(attr_model)
+            name_model_map[attr_name] = attr_model
+        elif attr_name in name_model_map:
+            conflicts.add(attr_name)
+            conflict_attrs = name_model_map.pop(attr_name)
+            name_model_map[attr_name] = attr_model
+            name_model_map[get_long_model_name(conflict_attrs)] = conflict_attrs
+            name_model_map[get_long_model_name(attr_model)] = attr_model
```

Comment:
It seems the second model with the same name will be always stored in the dict with short name.
For example: we insert 3 models A1,A2 and A3 with the short name `a`, and long name `long_a1`, `long_a2` and `long_a3` respectively:

1. Stored: `{a: A1}`
2. Pop `A1` and insert `A2`, then insert `A1` and `A2` again with long names, now the dict is `{a: A2, long_a1: A1, long_a2: A2}`
3. Insert `A3` with long name: `{a: A2, long_a1: A1, long_a2: A2, long_a3: A3}`

Since the set has undeterminstic order, so which one will be stored with the short name is in fact random. Therefore this implementation(popup and re-insert) looks unnecessarily complex to me.

Reply:
I think your suggestion makes sense.

I was using the same logic that pydantic used here.